✨ Add AsyncIPCProvider

dzhalaevd opened the pull request:
### What was wrong?

IPCProvider needs async version

Related to Issue https://github.com/ethereum/web3.py/issues/2973
Closes #1413 


### How was it fixed?
Add AsyncIPCProvider

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![image](https://github.com/ethereum/web3.py/assets/72649244/d35a5af5-f07a-41a5-a26f-5e32dc7e421e)

dzhalaevd committed:
✨ Add AsyncIPCProvider

dzhalaevd committed:
⚡️Add support for asynchronous context managers to Timeout

dzhalaevd committed:
✅ Update tests

dzhalaevd commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,155 @@
+        return self.sock
+
+
+class AsyncIPCProvider(JSONBaseProvider):
```
In the `AsyncIPCProvider` class, there is a synchronous context manager. However, the `_make_request` method is asynchronous. Therefore, I added support for asynchronous context managers to the `Timeout` class as part of the current task.

dzhalaevd commented on web3/_utils/threads.py:
```diff
@@ -99,7 +113,7 @@ def sleep(self, seconds: float) -> None:
 
     async def async_sleep(self, seconds: float) -> None:
-        await asyncio.sleep(seconds)
+        await asyncio.wait_for(asyncio.sleep(seconds), timeout=self.seconds)
```
The implementation of the `Timeout.async_sleep` method was modified to improve performance, if the delay exceeds the specified time, an exception will be thrown. This approach prevents program freezes in the event of a delay taking too long, providing improved reliability and responsiveness. However, I'm not sure if this is necessary in `async_sleep` or we can do with the old option.

dzhalaevd commented on web3/providers/async_ipc.py:
```diff
@@ -133,12 +136,12 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
 
             raw_response = b""
-            async with Timeout(self.timeout) as timeout:
+            async with asyncio.timeout(self.timeout) as timeout:
```
Why did you replace Timeout from `web3._utils.threads` on Timeout from `asyncio`?

dzhalaevd commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,158 @@
+            except BrokenPipeError:
+                # one extra attempt, then give up
+                sock = self._socket.reset()
+                sock.sendall(request)
```
```suggestion
                await sock.send(request)
```

dzhalaevd commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,158 @@
+                        else:
+                            return response
+                    else:
+                        await timeout.async_sleep(0)
```
```suggestion
                        await asyncio.sleep(0)
                    elif has_valid_json_rpc_ending(raw_response):
                        try:
                            response = self.decode_rpc_response(raw_response)
                        except JSONDecodeError:
                            await asyncio.sleep(0)
                            continue
                        else:
                            return response
                    else:
                        await asyncio.sleep(0)
```

JustAnotherC0der commented:
Pardon my ignorance to github work flow but is this implemented yet? Can't seem to import AsyncIPCProvider on newest web3

kclowes commented:
@JustAnotherC0der It's not yet. We're still working on it.

kclowes committed:
Add newsfragment

kclowes commented:
This is ready for review (finally)! I will add docs tomorrow, but wanted to get the code out there for feedback!

kclowes committed:
Add docs for AsyncIPCProvider, fix docs for IPCProvider

SheldonHolmgren commented:
Thanks for your hard work. I will really appreciate this feature.

Any plans to make AsyncIPCProvider a PersistentConnectionProvider? I think without this it will be impossible to `w3.eth.subscribe()`:
<details>

```

Web3ValidationError                       Traceback (most recent call last)
<ipython-input-12-2fc2efbc61b6> in <module>
----> 1 async with AsyncWeb3.persistent_websocket(AsyncIPCProvider("/mnt/ssd/optimism_data/geth.ipc")) as w3:
      2     await w3.eth.subscribe("newHeads")

~/dev/web3.py/web3/main.py in persistent_websocket(provider, middlewares, modules, external_modules, ens)
    516         a ``PersistentConnectionProvider`` instance.
    517         """
--> 518         return _PersistentConnectionWeb3(
    519             provider,
    520             middlewares,

~/dev/web3.py/web3/main.py in __init__(self, provider, middlewares, modules, external_modules, ens)
    541     ) -> None:
    542         if not isinstance(provider, PersistentConnectionProvider):
--> 543             raise Web3ValidationError(
    544                 "Provider must inherit from PersistentConnectionProvider class."
```
</details>

This is not a big deal since I can still use filters to get similar functionality.

pacrob reviewed (approved):
lgtm!

fselmo commented on web3/tools/benchmark/main.py:
```diff
@@ -139,6 +151,7 @@ def main(logger: logging.Logger, num_calls: int) -> None:
                     "params": {},
                     "exec": lambda: w3_http.eth.gas_price,
                     "async_exec": lambda: async_w3_http.eth.gas_price,
+                    "ipc": lambda: w3_ipc.eth.gas_price,
```
Did we want to add `async_ipc` here as well?

kclowes replied:
Eventually yes, but currently there is some bug where the `async_ipc` `send_transaction` stalls out at 41 calls via the benchmarking suite. I can make 500 `send_transaction` calls attached to a local dev node super fast, but haven't figured out the problem with benchmarking yet. I didn't want to spend too much time, so decided to add sync ipc calls and we can add async later if we want in a separate PR.

fselmo replied:
Ah, I see. I vaguely remember some chatter about that. Thanks for the info 👍🏼

fselmo commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,166 @@
+        return await asyncio.open_unix_connection(ipc_path)
+
+
+class PersistantSocket:
```
persistant -> persistent

```suggestion
class PersistentSocket:
```

I think this is copied from the `PersistantSocket` defined in the synchronous `IPCProvider`. Perhaps `AsyncPersistentSocket` (or `PersistentAsyncSocket`?) might be best in case it's ever imported anywhere then we don't have the same class name imported from two separate modules? Maybe we can fix the spelling on that one in `v7` too 😅 .

kclowes replied:
Ha thanks. Yes it is 😆

fselmo commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,166 @@
+                    writer.write(request)
+                    await writer.drain()
+
+            async def decode_response() -> RPCResponse:
```
nit: If we define this method outside of the scope of the async context manager and call it from within the context instead, this might have a bit better performance as well as readability. I believe there's an overhead to redefining the method each time the async context manager is entered. Even if it's slight, it might read a bit better too? Thoughts?

Something like 

```py
async def get_response_from_reader(self, reader: asyncio.StreamReader) -> RPCResponse:
    raw_response = b""
    
    while True:
        ...
        
        return response
```

kclowes replied:
Done! Good call

fselmo commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,166 @@
+        return self.sock
+
+
+_async_session_cache_lock = threading.Lock()
```
Is there a session or a session cache that the IPC Provider is utilizing? If we don't plan on doing anything related to any sessions and if it's not sharing anything across provider instances, I wonder if we can't make use of an `asyncio.Lock` by itself here.

We currently use one [here](https://github.com/ethereum/web3.py/blob/main/web3/providers/persistent.py#L32) on the `PersistentConnectionProvider` to keep separate parts of the code base from calling `recv()` on the websocket connection at the same time. Though there's an open PR that would get rid of our need for that lock, it may help here. Thoughts?

kclowes replied:
Oop, yep. I think an `asyncio.Lock` should work by itself. Nice catch!

fselmo commented on web3/providers/async_ipc.py:
```diff
@@ -0,0 +1,166 @@
+    def __str__(self) -> str:
+        return f"<{self.__class__.__name__} {self.ipc_path}>"
+
+    async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
```
This question builds a bit on the comment from @SheldonHolmgren. ~~If we create a reader and writer here each time we make a request, we aren't utilizing the "persistent open socket" paradigm in a way that could be advantageous to state preservation requests like subscriptions.~~ 

[Nope, it's staying alive. User error / I wasn't testing correctly. I stand corrected. I wonder how much it would take then to build subscription support, even if it doesn't inherit from `PersistentConnectionProvider` too well. It may take more work and breaking changes to try to come back later and put that work in. It may take just utilizing the reader in an asynchronous iterator context and trying to make use of the `_PersistentConnectionWeb3` class? 🤔 ]

~~I'm curious how much work it may take to make this a `PersistentConnectionProvider` or something similar to where we can keep an open connection throughout the life of the provider rather than opening / closing each time.~~ 

Not necessarily trying to head in that direction as that may take quite a bit more work, but curious on everyone's thoughts on that.

---
edited to reflect only how we may inherit from `PersistentConnectionProvider` and / or add `eth_subscribe` support 🤔

kclowes replied:
Yeah, I was thinking about that too. IMHO it makes sense to allow for `eth_subscribe` support here, but not sure how much work it'll take. I'll timebox it and if I can get something working, I'll add in a new commit for reviewing.

fselmo replied:
Sounds good. In a best case scenario, assuming we go with the message listener task that's currently in PR for the `WebsocketProviderV2`, I think we can probably hijack all of the `RequestProcessor` logic, switch the naming within the more broad `PersistentConnectionProvider` class and the `RequestManager` (from `ws` or `websocket` to `socket` maybe?) while using the appropriate logic for the specific sockets in the implementation classes. We can likely route the `socket.reader` into a message listener background task in a similar way as iterating over new messages on the websocket is done in wsv2. Curious to see what you find out as you timebox it though 👀

fselmo reviewed (commented):
Thanks for everyone's contribution here. I think this will be a great option to have. I left some initial comments. Mostly curious about the `PersistentConnectionProvider` question.

fselmo committed:
Inherit from PersistentConnectionProvider for AsyncIPCProvider

- Iron out some issues with rebasing; update async_ipc

kclowes committed:
AsyncIPCProvider updates:

- Time subscription tests correctly
- Fix lint
- Fix doc underline
- Remove sleeps in test to test multiple responses coming in quick succession
- Handle when multiple responses come in at once
- Add test for subscription
- timeout -> request_timeout

Updates to naming:
  - Remove ws specific naming from manager methods
  - _ws_message_listener -> _message_listener
  - WebsocketConnection -> PersistentConnection
  - persistent_websocket -> persistent_connection

SheldonHolmgren commented:
This code won't work if the client receives multiple responses before it reads the first one. I have coded up a dirty fix for my own use. Perhaps it might be useful as a hint: https://github.com/DavidRomanovizc/web3.py/compare/main...SheldonHolmgren:async_ipc_multiple_resps

kclowes commented:
Thank you @SheldonHolmgren! I'll see what I can do.

fselmo committed:
Clean up ``AsyncIPCProvider``; move some settings to base class:

- Clean up ``AsyncIPCProvider``. Some logging c/p was left in the code.
- Remove the need for the ``PersistentSocket`` class.
- Move the ``silence_listener_task_exceptions`` setting to the base class,
  ``PersistentConnectionProvider``.
- Use the suggestion put forth in the PR #2984 for parsing messages from the
  IPC socket more efficiently in case more than one message is received at
  once.

fselmo committed:
Clean up + fix AsyncIPC tests; move fixtures and constants to top of file

fselmo committed:
Ensure the listen event is cleared before waiting until set

fselmo committed:
Refactor wsv2 and async_ipc into providers/persistent

- Update docs for persistent connection provider refactor

Remove sleeps to emulate responses in rapid succession

kclowes commented:
All right, I think this is ready again for another look. Thanks @fselmo for the boost!

fselmo reviewed (approved):
I caught one minor doc change we missed but that was all I saw (and I went ahead and pushed it). Thanks for all the work here @DavidRomanovizc, @kclowes, and @SheldonHolmgren! lgtm 👍🏼 

@kclowes I'll let you merge this in when you are good with it. Maybe some more squashing if it makes sense too?

pacrob commented on web3/manager.py:
```diff
@@ -314,25 +314,25 @@ async def coro_request(
+    async def _message_stream(self) -> AsyncGenerator[RPCResponse, None]:
         if not isinstance(self._provider, PersistentConnectionProvider):
             raise TypeError(
                 "Only websocket providers that maintain an open, persistent connection "
```
It looks like we're removing references to websockets, and either just using "socket" or "persistent connection". Should that happen here too?

kclowes replied:
oop, yep. Nice catch!

kclowes committed:
clean up websocket verbiage

kclowes committed:
Update docs, minor fixes

pacrob commented on docs/providers.rst:
```diff
@@ -261,7 +376,8 @@ Persistent Connection Providers
+    and the ``AsyncIPCProvider``.
     It handles interactions with a persistent connection to a JSON-RPC server. Among
     its configuration, it houses all of the
     :class:`~web3.providers.websocket.request_processor.RequestProcessor` logic for
```
This moved to `web3.providers.persistent.request_processor` I believe.
